File: script/resourceManager/Shader.js
define([],
    function () {
        "use strict";

        /**
         *
         * @constructor
         * @param {{id:string, vertex:string, fragment:string}} jShader Shader JSON object
         * @param {string} sVertexShader Vertex shader as source text
         * @param {string} sFragmentShader Fragment shader as source text
         * @param gl OpenGL context object
         */
        var Shader = function (jShader, sVertexShader, sFragmentShader, gl) {
            this.id = jShader.id;
            this.gl = gl;

            this.glVertexShader = gl.createShader(gl.VERTEX_SHADER);
            this.glFragmentShader = gl.createShader(gl.FRAGMENT_SHADER);

            gl.shaderSource(this.glVertexShader, sVertexShader);
            gl.compileShader(this.glVertexShader);
            if (!gl.getShaderParameter(this.glVertexShader, gl.COMPILE_STATUS)) {
                window.alert(gl.getShaderInfoLog(this.glVertexShader));
            }

            gl.shaderSource(this.glFragmentShader, sFragmentShader);
            gl.compileShader(this.glFragmentShader);
            if (!gl.getShaderParameter(this.glFragmentShader, gl.COMPILE_STATUS)) {
                window.alert(gl.getShaderInfoLog(this.glFragmentShader));
            }

            this.glProgram = gl.createProgram();
            gl.attachShader(this.glProgram, this.glVertexShader);
            gl.attachShader(this.glProgram, this.glFragmentShader);
            gl.linkProgram(this.glProgram);
            if (!gl.getProgramParameter(this.glProgram, gl.LINK_STATUS)) {
                window.alert("Could not initialise shaders");
            }

            this.glVertexPositionAttribute = gl.getAttribLocation(this.glProgram, "aVertexPosition");
            gl.enableVertexAttribArray(this.glVertexPositionAttribute);
            this.glTextureCoordinateAttribute = gl.getAttribLocation(this.glProgram, "aTextureCoordinate");
            gl.enableVertexAttribArray(this.glTextureCoordinateAttribute);
            this.glTextureUniform = gl.getUniformLocation(this.glProgram, "uTexture");

            this.mProjection = gl.getUniformLocation(this.glProgram, "uPMatrix");
            this.mModelView = gl.getUniformLocation(this.glProgram, "uMVMatrix");

        };

        Shader.prototype.use = function (mProjection) {
            this.gl.useProgram(this.glProgram);
            this.gl.uniformMatrix4fv(this.mProjection, false, mProjection);
        };

        return Shader;
    }
);

File: lib/resource_management/shader.dart
part of resource_management;

class Shader extends Resource {

  String id;

  gl.RenderingContext context;

  gl.Shader vertexShader;
  gl.Shader fragmentShader;
  gl.Program program;

  int vertexPositionAttribute;
  int textureCoordinateAttribute;
  gl.UniformLocation textureUniform;

  gl.UniformLocation projectionUniform;
  gl.UniformLocation modelViewUniform;

  Shader(Map shaderJSON, String vertexShader, String fragmentShader,
         gl.RenderingContext context) :
  this.context = context,
  this.vertexShader = context.createShader(gl.VERTEX_SHADER),
  this.fragmentShader = context.createShader(gl.FRAGMENT_SHADER),
  this.program = context.createProgram(),
  super(shaderJSON['id']) {

    context.shaderSource(this.vertexShader, vertexShader);
    context.compileShader(this.vertexShader);
    if (!context.getShaderParameter(this.vertexShader, gl.COMPILE_STATUS)) {
      window.alert(context.getShaderInfoLog(this.vertexShader));
    }

    context.shaderSource(this.fragmentShader, fragmentShader);
    context.compileShader(this.fragmentShader);
    if (!context.getShaderParameter(this.fragmentShader, gl.COMPILE_STATUS)) {
      window.alert(context.getShaderInfoLog(this.fragmentShader));
    }

    context.attachShader(this.program, this.vertexShader);
    context.attachShader(this.program, this.fragmentShader);
    context.linkProgram(this.program);
    if (!context.getProgramParameter(this.program, gl.LINK_STATUS)) {
      window.alert('Could not link program');
    }

    this.vertexPositionAttribute = context.getAttribLocation(this.program,
    'aVertexPosition');
    context.enableVertexAttribArray(this.vertexPositionAttribute);
    this.textureCoordinateAttribute = context.getAttribLocation(this.program,
    'aTextureCoordinate');
    context.enableVertexAttribArray(this.textureCoordinateAttribute);
    this.textureUniform = context.getUniformLocation(this.program, 'uTexture');


    this.projectionUniform = context.getUniformLocation(this.program,
    'uPMatrix');
    this.modelViewUniform = context.getUniformLocation(this.program,
    'uMVMatrix');
  }

  void use(Matrix4 projectionMatrix) {
    this.context.useProgram(this.program);
    this.context.uniformMatrix4fv(this.projectionUniform, false,
    projectionMatrix.storage);
  }
}